### PR TITLE
feat(pageviews): enable filters for checkpoint, source, and target

### PR DESCRIPTION
**Pull request by trieloff**
This is a massive refactoring, and simplification, borrowing the filtering from #1053

It drops the support for `granularity=auto` so we probably have to declare this a breaking change, even if the impact is much smaller than the last breaking change we had.

It also partially reverts the changes made in #1039, negative offsets are now possible again, so that you can use `startdate` and `enddate` parameters

### CODE REVIEW

**Comment by langswei** [comment]
From reading, not yet testing -- the UNION ALL technique as a replacement for CASE is quite clever.  This is easier to read than the previous version, perhaps not quite as efficient given the multiple EVENTS_V3 table function calls but likely still very fast.

I'm not confident that the TIMESTAMP_TRUNC with a default or non-negative offset+interval will give us what we want but will reserve judgment until after testing.  For example, I imagine a mid-February call of `granularity=30&interval=3` with default `offset=0` would return half of November, all of December, all of January, and half of February.  Maybe there is no more sensible behavior and the user will just learn to work with it.

An automated test fails.  If you don't get to this during your day, I can check it out during my day.

File: src/queries/rum-pageviews.sql [hunk 1]
```diff
@@ -1,137 +1,250 @@
...
+  SELECT
+    id,
+    checkpoint,
+    source,
+    target,
+    pageviews,
+    TIMESTAMP_TRUNC(time, ISOWEEK) AS trunc_date
+  FROM helix_rum.EVENTS_V3(
+    @url, # url
+    @offset * 7, # offset in weeks
```
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:45]: This needs to be cast to int64 as Events_V3  expects in64, int64 for those two in the signature.

File: src/queries/rum-pageviews.sql [hunk 2]
```diff
@@ -1,137 +1,250 @@
...
+    @interval * 7, # weeks to fetch
```
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:46]: this also needs to be cast to int64

File: src/queries/rum-pageviews.sql [hunk 3]
```diff
@@ -1,137 +1,250 @@
...
+  SELECT
+    id,
+    checkpoint,
+    source,
+    target,
+    pageviews,
+    TIMESTAMP_TRUNC(time, MONTH) AS trunc_date
+  FROM helix_rum.EVENTS_V3(
+    @url, # url
+    @offset * 30, # offset in months
```
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:65]: int64

File: src/queries/rum-pageviews.sql [hunk 4]
```diff
@@ -1,137 +1,250 @@
...
+    @interval * 30, # months to fetch
```
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:66]: int64

File: src/queries/rum-pageviews.sql [hunk 5]
```diff
@@ -1,137 +1,250 @@
...
+  SELECT
+    id,
+    checkpoint,
+    source,
+    target,
+    pageviews,
+    TIMESTAMP_TRUNC(time, QUARTER) AS trunc_date
+  FROM helix_rum.EVENTS_V3(
+    @url, # url
+    @offset * 90, # offset in quarters
```
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:85]: cast to int64

File: src/queries/rum-pageviews.sql [hunk 6]
```diff
@@ -1,137 +1,250 @@
...
+    @interval * 90, # quarters to fetch
```
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:86]: cast to int64

File: src/queries/rum-pageviews.sql [hunk 7]
```diff
@@ -1,137 +1,250 @@
...
+  SELECT
+    id,
+    checkpoint,
+    source,
+    target,
+    pageviews,
+    TIMESTAMP_TRUNC(time, YEAR) AS trunc_date
+  FROM helix_rum.EVENTS_V3(
+    @url, # url
+    @offset * 365, # offset in years
```
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:105]: cast as int64

File: src/queries/rum-pageviews.sql [hunk 8]
```diff
@@ -1,137 +1,250 @@
...
+    @interval * 365, # years to fetch
```
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:106]: cast as int64

File: src/queries/rum-pageviews.sql [hunk 9]
```diff
@@ -1,137 +1,250 @@
...
+  SELECT
+    id,
+    checkpoint,
+    source,
+    target,
+    pageviews,
+    TIMESTAMP_TRUNC(time, DAY) AS trunc_date
+  FROM helix_rum.EVENTS_V3(
+    @url, # url
+    @offset, # offset
```
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:25]:
```suggestion
    CAST(@offset as INT64), # offset
```

**Review by MarquiseRosier — CHANGES_REQUESTED** [review]

File: src/queries/rum-pageviews.sql [hunk 10]
```diff
@@ -1,137 +1,250 @@
...
+    @interval, # days to fetch
```
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:26]:
```suggestion
    CAST(@interval as INT64), # days to fetch
```

File: src/queries/rum-pageviews.sql (hunk 1 above)
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:45]:
```suggestion
    CAST(@offset AS INT64) * 7, # offset in weeks
```

File: src/queries/rum-pageviews.sql (hunk 2 above)
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:46]:
```suggestion
    CAST(@interval AS INT64) * 7, # weeks to fetch
```

File: src/queries/rum-pageviews.sql (hunk 3 above)
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:65]:
```suggestion
    CAST(@offset AS INT64) * 30, # offset in months
```

File: src/queries/rum-pageviews.sql (hunk 4 above)
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:66]:
```suggestion
    CAST(@interval AS INT64) * 30, # months to fetch
```

File: src/queries/rum-pageviews.sql (hunk 5 above)
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:85]:
```suggestion
    CAST(@offset AS INT64) * 90, # offset in quarters
```

File: src/queries/rum-pageviews.sql (hunk 6 above)
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:86]:
```suggestion
    CAST(@interval AS INT64) * 90, # quarters to fetch
```

File: src/queries/rum-pageviews.sql (hunk 7 above)
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:105]:
```suggestion
    CAST(@offset AS INT64) * 365, # offset in years
```

File: src/queries/rum-pageviews.sql (hunk 8 above)
Comment (MarquiseRosier) [inline src/queries/rum-pageviews.sql:106]:
```suggestion
    CAST(@interval AS INT64) * 365, # years to fetch
```

**Review by MarquiseRosier — CHANGES_REQUESTED** [review]
I think these need to be casted to INT64 before the multiplication! :)

**Comment by trieloff** [comment]
> perhaps not quite as efficient given the multiple EVENTS_V3 table function calls but likely still very fast.
The query optimizer seems to be clever enough not to expand the CTE before evaluating the `WHERE` statement, so we are not overfetching

**Review by MarquiseRosier — APPROVED** [review]

**Pull request merged**